### PR TITLE
Update Template - Metadata/AM Mapping

### PR DESCRIPTION
**Pull request by TayshawnH**
## 💸 TL;DR

This PR updates the GTM template for collecting Google Analytics (GA4) event data, specifically metadata and match keys, and mapping them to Reddit Pixel events. 

## 📜 Details
[Design Doc](https://docs.google.com/document/d/1W7hr4IL3tqCEoOpyVq6VOKwEdV2mBQ40sXTreJTJTsw/edit?tab=t.0#heading=h.ymk2exkyyv60)

[Jira](https://reddit.atlassian.net/browse/ACS-2966)

### GA4 `ecommerce` Mapping:
- Collects and processes `ecommerce.items` array (from `ecommerce` or `eventModel` objects) into `eventMetadata.products`
- Concatenating hierarchical categories (item_category, item_category2, etc.) into product.category field, separated by `>`
- Calculates `eventMetadata.itemCount` by summing `item.quantity` from the `ecommerce.items` array

### GA4 `user_data` Mapping:
- Maps email_address/sha256_email_address to the email match key.
- Maps phone_number/sha256_phone_number to the phoneNumber match key.

## 🧪 Testing Steps / Validation

- Existing tests passing
- New tests added
- Tested via test page and Reddit Pixel Helper

dataLayer.push from test page - tayshawnhill.dev:
```
window.dataLayer.push(
{
    "event": "add_to_cart",
    "user_data": {
        "email_address": "alice@site.com",
        "phone_number": "+15551234567"
    },
    "ecommerce": {
        "currency": "USD",
        "value": 19.99,
        "items": [
            {
                "item_id": "SKU_1",
                "item_name": "T-Shirt",
                "price": 19.99,
                "quantity": 1,
                "item_category": "Apparel",
                "item_category2": "Mens",
                "item_category3": "Shirts"
            },
            {
                "item_id": "SKU_123",
                "item_name": "Standard T-Shirt",
                "item_category": "Apparel",
                "price": 25,
                "quantity": 2
            },
            {
                "item_id": "SKU_456",
                "item_name": "Hiking Boots",
                "item_category": "Shoes",
                "item_category2": "Outdoor",
                "item_category3": "Hiking",
                "price": 50,
                "quantity": 1
            },
            {
                "item_id": "ALTID_789",
                "item_name": "watch",
                "item_category": "Accessories",
                "price": 15,
                "quantity": 10
            },
            {
                "item_id": "SKU_CAT_EDGE",
                "item_name": "Category Test Item",
                "item_category": " Home Goods  ",
                "item_category3": "Kitchen",
                "price": 10,
                "quantity": 3
            }
        ]
    },
})
```

rp.gif response payload:
```
ts: 1744735080362
id: a2_eztpxdd5mulb
event: add_to_cart
m.itemCount: 17
m.value: 
m.valueDecimal: 19.99
m.currency: USD
m.transactionId: 
m.customEventName: 
m.products: [{"id":"SKU_1","name":"T-Shirt","category":"Apparel > Mens > Shirts"},{"id":"SKU_123","name":"Standard T-Shirt","category":"Apparel"},{"id":"SKU_456","name":"Hiking Boots","category":"Shoes > Outdoor > Hiking"},{"id":"ALTID_789","name":"watch","category":"Accessories"},{"id":"SKU_CAT_EDGE","name":"Category Test Item","category":"Home Goods > Kitchen"}]
m.conversionId: 
uuid: 8f51cbe8-ea00-4eec-903b-467868dbd389
aaid: 
em: 825f7f3f63efec16f0230fbc37e80a0965f591c6926d0c7e77dd34ae5d4507d8
pn: 8a59780bb8cd2ba022bfa5ba2ea3b6e07af17a7d8b30c1f9b3390e36f69019e4
external_id: 
idfa: 
integration: gtm
partner: AUTOMATIC_GTM
opt_out: 0
sh: 5120
sw: 1440
v: rdt_d9500dd4
dpm: 
dpcc: 
dprc: 
```

### Regression Check
We need to make sure core functionality for users _not_ utilizing GA4 `ecommerce`/`user_data` features should remain unaffected. 

## ✅ Checks

- [ ] CI tests (if present) are passing
- [ ] Adheres to code style for repo
- [ ] Contributor License Agreement (CLA) completed if not a Reddit employee

### CODE REVIEW

File: template.tpl [hunk 1]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
+} else {
+  eventMetadata.currency = data.currency || copyFromDataLayer("currency") || (eventModel && eventModel.currency);
+}
 
-// If product metadata is available in the data layer, use it
-if (productRowsFromDataLayer && productRowsFromDataLayer.length) {
-  eventMetadata.products = productRowsFromDataLayer;
-} else if (productJSONFromDataLayer && productJSONFromDataLayer.length) {
-  eventMetadata.products = productJSONFromDataLayer;
+if (ecommerce && ecommerce.value !== undefined) {
```
[the remark this answers is not in the document]
Comment (TayshawnH) [reply template.tpl:580]: yeah that is right. thanks for catching

File: template.tpl [hunk 2]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
-eventMetadata.value = data.transactionValue || copyFromDataLayer("transactionValue") || (eventModel && eventModel.transactionValue);
-
-// Try grabbing product metadata from dataLayer
-const productRowsFromDataLayer = copyFromDataLayer("productsRows") || (eventModel && eventModel.productsRows);
-const productJSONFromDataLayer = copyFromDataLayer("productsJSON") || (eventModel && eventModel.productsJSON);
+if (ecommerce && ecommerce.currency) {
+  eventMetadata.currency = ecommerce.currency;
+} else {
+  eventMetadata.currency = data.currency || copyFromDataLayer("currency") || (eventModel && eventModel.currency);
+}
```
Comment (eseosaasiruwa) [inline template.tpl:578]: i think the ordering here might be off? it seems like we're prioritizing currency found in the ecommerce struct instead of the UI or datalayer. For example, for phoneNumber and email above, we're only setting those with the data found in userData if they're not already set

File: template.tpl [hunk 3]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
-    // We can assign it directly.
-    eventMetadata.products = data.productsRows;
-  } else if (data.productInputType == "entryJSON" && data.productsJSON && data.productsJSON.length) {
-    // The text input should be stringified JSON. The Pixel handles validation/normalization.
-    // We can assign it directly.
-    eventMetadata.products = data.productsJSON;
+  eventMetadata.value = data.transactionValue || copyFromDataLayer("transactionValue") || (eventModel && eventModel.transactionValue);
+}
+
+var formatCategories = function (item) {
```
Comment (eseosaasiruwa) [inline template.tpl:586]: we should have a unit test for this function

File: template.tpl [hunk 4]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
+      if (value !== "") {
+        categories.push(value);
+      }
+    }
   }
+  
+  return categories.length > 0 ? categories.join(" > ") : null;
+};
+
+var processEcommerceItems = function () {
```
Comment (eseosaasiruwa) [inline template.tpl:603]: same note about having a unit test for this function

File: template.tpl [hunk 5]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
+        itemCount: processedItemCount,
+        foundItems: true,
+      };
+    }
+  }
+  // Return default if no items found, array was empty, or no valid products resulted
+  return { products: [], itemCount: 0, foundItems: false };
+};
+
+var getProductData = function () {
```
Comment (eseosaasiruwa) [inline template.tpl:644]: same note about unit test

**Review by eseosaasiruwa — CHANGES_REQUESTED** [review]

File: template.tpl [hunk 6]
```diff
@@ -558,38 +571,166 @@ if (!enableFirstPartyCookies) {
...
-  // If there is no product metadata in the data layer, fall back to checking productInputType
-  if (data.productInputType == "entryManual" && data.productsRows && data.productsRows.length) {
-    // The simple table is of the correct format - an array of objects.
-    // We can assign it directly.
-    eventMetadata.products = data.productsRows;
-  } else if (data.productInputType == "entryJSON" && data.productsJSON && data.productsJSON.length) {
-    // The text input should be stringified JSON. The Pixel handles validation/normalization.
-    // We can assign it directly.
-    eventMetadata.products = data.productsJSON;
+  eventMetadata.value = data.transactionValue || copyFromDataLayer("transactionValue") || (eventModel && eventModel.transactionValue);
```
[the remark this answers is not in the document]
Comment (TayshawnH) [reply template.tpl]: I updated the template to check for UI > data layer

File: template.tpl (hunk 2 above)
Comment (TayshawnH) [reply template.tpl]: updated the priority order of the fields

File: template.tpl (hunk 3 above)
Comment (TayshawnH) [reply template.tpl:577]: I approached testing from an integration/outcome perspective rather than directly unit testing the helper functions. The GTM template testing env has limitation when trying to directly access internal function and I could not find a good way without just mocking it. I just created test to verify the outcomes of these functions. 

If needed, we can add more comprehensive ones than we currently have.

File: template.tpl [hunk 7]
```diff
@@ -558,38 +571,142 @@ if (!enableFirstPartyCookies) {
   _rdt('disableFirstPartyCookies');
 }
 
-eventMetadata.currency = data.currency || copyFromDataLayer("currency") || (eventModel && eventModel.currency);
-eventMetadata.value = data.transactionValue || copyFromDataLayer("transactionValue") || (eventModel && eventModel.transactionValue);
+eventMetadata.currency = data.currency || copyFromDataLayer("currency") || (eventModel && eventModel.currency) || (ecommerce && ecommerce.currency);
+eventMetadata.value = data.transactionValue || copyFromDataLayer("transactionValue") || (eventModel && eventModel.transactionValue) || (ecommerce && ecommerce.value);
```
Comment (TayshawnH) [inline template.tpl:575]: Can value be `0`? I am thinking of the case of where users want to explicitly set something to zero (for like free items or promotions). if they send use zero for the datalayer it would get skipped over since it would treat it as falsy? @eseosaasiruwa @creativefunction13

Comment (creativefunction13) [reply template.tpl:575]: Yeah, we should accept value `0`.  It's tough because sometimes that intentional as you say, and other times it's unintentional because a bad implementation is using a non-nullable number for value.  Either way we should accept it.

Comment (TayshawnH) [reply template.tpl]: updated here - 4aca9db7afbf1bbc0d826fcfc5184441c6137c9b

**Review by creativefunction13 — APPROVED** [review]

File: template.tpl (hunk 3 above)
Comment (eseosaasiruwa) [reply template.tpl:592]: I would lean towards having comprehensive unit tests but I wont block on that

**Review by eseosaasiruwa — APPROVED** [review]
remember, we don't want to merge this yet. work with Austin on testing all edge cases first. once everything looks good, we can follow up with PMM and PM about release

File: template.tpl [hunk 8]
```diff
@@ -561,6 +578,20 @@ if (dataProcessingOptions.region) {
...
+  // Check for traditional Tag Manager API
+  if (ecommerce || userData) dataSource += 1;
+  // Check for Google Tag API
+  if (eventModel && (eventModel.user_data || eventModel.items || eventModel.currency || eventModel.value)) dataSource += 2;
+
+  // if some how both set data source to 3
+  return dataSource.toString();
+};
+
+initData.partner_version = "GTM_" + templateVersion + ":" + getUsageProfile();
```
Comment (creativefunction13) [inline template.tpl:593]: Is pixel going to use this information?  We probably need updates all the way into the system with this right?

Comment (TayshawnH) [reply template.tpl:593]: Yeah we will use this downstream - https://docs.google.com/document/d/1gAofwCbbDGbiXgeoEELH7ONgpl_dz9Z7zGqPjtv9eew/edit?tab=t.0#heading=h.ytpri06w367x
